pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,295 @@
+    _logger = LoggerFactory.getLogger(tableNameWithType + "-" + getClass().getSimpleName());
+    _serverMetrics = serverMetrics;
+    _tableNameWithType = tableNameWithType;
+    _realTimeTableDataManager = realtimeTableDataManager;
+    // Handle negative timer values
+    if ((initialTimeThreadDelayMs < 0) || (timerThreadTickIntervalMs <= 0)) {
+      throw new RuntimeException("Illegal timer arguments");
+    }
+    _enableAging = true;
+    _initialTimeThreadDelayMs = initialTimeThreadDelayMs;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timer = new Timer("ConsumptionDelayTimerThread" + tableNameWithType);
+    _timer.schedule(new TrackingTimerTask("TimerTask" + tableNameWithType, this),
+        _initialTimeThreadDelayMs, _timerThreadTickIntervalMs);
+    // Install callback metric
+    _serverMetrics.addCallbackTableGaugeIfNeeded(_tableNameWithType, ServerGauge.MAX_PINOT_CONSUMPTION_DELAY_MS,
+        () -> (long) getMaxPinotConsumptionDelay());
+  }
+
+  // Constructor that uses default timeout
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, tableNameWithType, tableDataManager, INITIAL_TIMER_THREAD_DELAY_MS,
+        TIMER_THREAD_TICK_INTERVAL_MS);
+  }
+
+  /**
+   * Use to set or rest the aging of reported values.
+   * @param enableAging true if we want maximum to be aged as per sample time or false if we do not want to age
+   *                   samples
+   */
+  @VisibleForTesting
+  void setEnableAging(boolean enableAging) {
+    _enableAging = enableAging;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   * If the new sample represents a new Maximum we update the current maximum.
+   * If the new sample was for the partition that was maximum, but delay is not maximum anymore, we must select
+   * a new maximum.
+   *
+   * @param delayInMilliseconds pinot consumption delay being recorded.
+   * @param sampleTime sample time.
+   * @param partitionGroupId partition ID for which this delay is being recorded.
+   */
+  public void storeConsumptionDelay(long delayInMilliseconds, long sampleTime, long partitionGroupId) {
+    // Store new measure and wipe old one for this partition
+    _partitionToDelaySampleMap.put(partitionGroupId, new DelayMeasure(sampleTime, delayInMilliseconds));
+    // If we are consuming we do not need to track this partition for removal.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+  }
+
+  /*
+   * Handle partition removal event. This must be invoked when we stop serving a given partition for
+   * this table in the current server.
+   * This function will be invoked when we receive CONSUMING -> DROPPED / OFFLINE state transitions.
+   *
+   * @param partitionGroupId partition id that we should stop tracking.
+   */
+  public void stopTrackingPartitionConsumptionDelay(long partitionGroupId) {
```

Comment:
Why is partitionGroupId long? We have it as int everywhere else

Reply:
Seems like I missed this in my push yesterday but it is in my dev env.